src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -9054,7 +9075,24 @@ private ExpressionSyntax ParseSubExpressionCore(Precedence precedence)
                     }
                     else
                     {
-                        leftOperand = _syntaxFactory.BinaryExpression(opKind, leftOperand, opToken, this.ParseSubExpression(newPrecedence));
+                        if (tk == SyntaxKind.DotDotToken)
+                        {
+                            // Operator ".." here can either be a binary or a postfix unary operator:
+                            Debug.Assert(opKind == SyntaxKind.RangeExpression);
+
+                            if (IsPossibleExpression(allowBinaryExpressions: false, allowAssignmentExpressions: false))
+                            {
+                                leftOperand = _syntaxFactory.RangeExpression(leftOperand: leftOperand, opToken, rightOperand: this.ParseSubExpression(newPrecedence));
+                            }
+                            else
+                            {
+                                leftOperand = _syntaxFactory.RangeExpression(leftOperand: leftOperand, opToken, rightOperand: null);
+                            }
```

Comment:
Consider parsing `rightOperand` of the range expression inside the `if/else` and call `_syntaxFactory.RangeExpression()` after the `if/else`.